src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs

```diff
@@ -166,86 +169,125 @@ private static Task<Solution> GetSolutionAsync(RoslynServices roslynService, Pin
 
+        [Obsolete("Use one with no CancellationToken given. underlying issue has been addressed", error: true)]
         protected async Task RunServiceAsync(Func<CancellationToken, Task> callAsync, CancellationToken cancellationToken)
         {
             AssetStorage.UpdateLastActivityTime();
 
-            // merge given cancellation token with shutdown cancellation token. it looks like if cancellation and disconnection happens
-            // almost same time, we might not get cancellation message back from stream json rpc and get disconnected
-            // https://github.com/Microsoft/vs-streamjsonrpc/issues/64
-            using (var mergedCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ShutdownCancellationToken))
+            try
             {
-                try
-                {
-                    await callAsync(mergedCancellation.Token).ConfigureAwait(false);
-                }
-                catch (Exception ex) when (LogUnlessCanceled(ex, mergedCancellation.Token))
-                {
-                    // never reach
-                    return;
-                }
+                await callAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex) when (LogUnlessCanceled(ex, cancellationToken))
+            {
+                // never reach
+                return;
             }
         }
 
+        [Obsolete("Use one with no CancellationToken given. underlying issue has been addressed", error: true)]
         protected T RunService<T>(Func<CancellationToken, T> call, CancellationToken cancellationToken)
         {
             AssetStorage.UpdateLastActivityTime();
 
-            // merge given cancellation token with shutdown cancellation token. it looks like if cancellation and disconnection happens
-            // almost same time, we might not get cancellation message back from stream json rpc and get disconnected
-            // https://github.com/Microsoft/vs-streamjsonrpc/issues/64
-            using (var mergedCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ShutdownCancellationToken))
+            try
             {
-                try
-                {
-                    return call(mergedCancellation.Token);
-                }
-                catch (Exception ex) when (LogUnlessCanceled(ex, mergedCancellation.Token))
-                {
-                    // never reach
-                    return default;
-                }
+                return call(cancellationToken);
+            }
+            catch (Exception ex) when (LogUnlessCanceled(ex, cancellationToken))
+            {
+                // never reach
+                return default;
             }
         }
 
+        [Obsolete("Use one with no CancellationToken given. underlying issue has been addressed", error: true)]
         protected void RunService(Action<CancellationToken> call, CancellationToken cancellationToken)
```

Comment:
canwe not just remove thse?

Reply:
not sure what you meant.

Reply:
Instead of Obsoleting these, why not just remove them entirely and update the callers.  This is all internal to Roslyn right?

Reply:
no it is used by partner team.

Reply:
why can the partners not move to the new API?

Reply:
I assume our pattern would be
1. Obsolete the API and insert
2. Let all partner teams consume the new binaries
3. Wait for sign-off that all partner teams have finished moving off the old API
4. Delete old API and insert

@heejaechang 	is this the plan?

Reply:
Oh, it's teams *outside* of Roslyn?  We're exposing our remote RPC system for other teams to use?  That seems... interesting :)   Could we abstract things away so they don't need to call into this sort of low level construct directly?  It seems like it introduces an artificial fragility point.

For example, CodeActions can run OOP, but don't need to have awareness of the RPC layer.

Reply:
Note: if it's teams outside of Roslyn then i retract my concern.  I can see why we'd have to go this route.

Reply:
many other teams are using our OOP for their feature. it is designed so that other team can use our OOP without in our repro from the beginning. not sure what is so interesting about it. and yes, @jmarolf it is following the user pattern.